### PR TITLE
[machine, node] make virtual app sequence numbers unique

### PR DESCRIPTION
**Pull request by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll**

### CODE REVIEW

File: packages/machine/src/models/state-channel.ts [hunk 1]
```diff
@@ -304,7 +304,8 @@ export class StateChannel {
       this.userNeuteredExtendedKeys,
       appInstances,
       this.ethVirtualAppAgreementInstances,
-      this.freeBalanceAppIndexes
+      this.freeBalanceAppIndexes,
+      this.monotonicNumInstalledApps + 1
```
Comment (snario) [inline packages/machine/src/models/state-channel.ts]: Don't think we need this + 1

File: packages/machine/src/protocol/install-virtual-app.ts [hunk 2]
```diff
@@ -133,13 +137,31 @@ export const INSTALL_VIRTUAL_APP_PROTOCOL: ProtocolExecutionFlow = {
...
+
+  const sc =
+    context.stateChannelsMap.get(key) ||
+    StateChannel.createEmptyChannel(key, [initiatingXpub, respondingXpub]);
+
+  const appSeqNo = sc.numInstalledApps;
+
+  const signingKeys = [xkeyKthAddress(intermediaryXpub, appSeqNo)].concat(
+    xkeysToSortedKthAddresses([initiatingXpub, respondingXpub], appSeqNo)
+  );
```
Comment (snario) [inline packages/machine/src/protocol/install-virtual-app.ts]: Why isn't the whole list sorted?

File: packages/machine/src/protocol/install-virtual-app.ts [hunk 3]
```diff
@@ -150,40 +172,25 @@ function createTarget(
...
-  const key = virtualChannelKey(
-    [initiatingXpub, respondingXpub],
-    intermediaryXpub
-  );
 
-  const sc = (
-    context.stateChannelsMap.get(key) ||
-    StateChannel.createEmptyChannel(key, [initiatingXpub, respondingXpub])
-  ).addVirtualAppInstance(targetAppInstance);
+  const newSc = sc.addVirtualAppInstance(target);
```
Comment (snario) [inline packages/machine/src/protocol/install-virtual-app.ts]: Can you add a check inside `addVirtualAppInstance` to verify the correct `appSeqNo` was set?

File: packages/machine/src/protocol/install-virtual-app.ts (hunk 2 above)
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/machine/src/protocol/install-virtual-app.ts]: https://github.com/counterfactual/monorepo/blob/master/packages/contracts/contracts/mixins/MixinVirtualAppSetState.sol#L99

Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/machine/src/protocol/install-virtual-app.ts]: https://github.com/counterfactual/monorepo/blob/master/packages/contracts/contracts/mixins/MixinVirtualAppSetState.sol#L19

File: packages/machine/src/models/state-channel.ts (hunk 1 above)
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/machine/src/models/state-channel.ts]: why not? we need the number to increment for the next time someone calls this function

Comment (snario) [reply packages/machine/src/models/state-channel.ts]: Ah, my bad, yeah

File: packages/machine/src/protocol/install-virtual-app.ts (hunk 2 above)
Comment (snario) [reply packages/machine/src/protocol/install-virtual-app.ts]: Got it, can you reference the specs in a comment in the code

**Pull request merged**